core/src/main/scala/cats/instances/map.scala

```diff
@@ -44,9 +44,21 @@ trait MapInstances extends cats.kernel.instances.MapInstances {
       )(f: A => G[B])(implicit G: CommutativeApplicative[G]): G[Map[K, B]] = {
         if (fa.isEmpty) G.pure(Map.empty[K, B])
         else
-          G.map(Chain.traverseViaChain(fa.toIndexedSeq) { case (k, a) =>
-            G.map(f(a))((k, _))
-          })(_.iterator.toMap)
+          G match {
+            case x: StackSafeMonad[G] =>
+              x.map(fa.foldLeft(G.pure(Map.newBuilder[K, B])) { case (accG, (k, a)) =>
```

Comment:
isn't this still mutable?

Reply:
ah sorry, I forgot that I special-cased this here. Will fix 👍

Reply:
this is still TODO it looks like

Reply:
This is still using the mutable builder, I think we want something like:
```scala
case x: StackSafeMonad[G] =>
  fa.iterator.foldLeft(G.pure(Map.empty[K, B])) { case (accG, (k, a)) =>
    x.map2(accG, f(a)) { case (acc, b) => acc + (k -> b) }
  }
```